File: 每日一题/5.8/3342.到达最后一个房间的最少时间-ii.cpp
/*
 * @lc app=leetcode.cn id=3342 lang=cpp
 *
 * [3342] 到达最后一个房间的最少时间 II
 */
#include "tools.h"
using namespace std;
// @lc code=start
class Solution {
public:
    int minTimeToReach(vector<vector<int>>& moveTime) {
        int m = moveTime.size(), n = moveTime[0].size();
        vector<vector<int>> dis(m, vector<int>(n, INT_MAX));
        dis[0][0] = 0;
        using pack = vector<int>;
        vector<vector<int>> d = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        priority_queue<pack, vector<pack>, greater<pack>> heap;
        heap.push({0, 0, 0});
        while (heap.size())
        {
            auto top = heap.top();
            int times = top[0], x = top[1], y = top[2];
            heap.pop();
            if (x == m - 1 && y == n - 1) return times;
            if (times > dis[x][y]) continue;
            for (int i = 0; i < d.size(); i ++ )
            {
                int nx = x + d[i][0], ny = y + d[i][1];
                if (nx >= 0 && nx < m && ny >= 0 && ny < n)
                {
                    int nxt_time = max(times, moveTime[nx][ny]) + (x + y) % 2 + 1;
                    if (nxt_time < dis[nx][ny])
                    {
                        dis[nx][ny] = nxt_time;
                        heap.emplace(vector<int>{nxt_time, nx, ny});
                    }
                }
            }
        }
        return -1;
    }
};
// @lc code=end

File: 每日一题/5.8/3342.到达最后一个房间的最少时间-ii.py
#
# @lc app=leetcode.cn id=3342 lang=python3
#
# [3342] 到达最后一个房间的最少时间 II
#
from mytools import *
# @lc code=start
class Solution:
    def minTimeToReach(self, g: List[List[int]]) -> int:
        d = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        m, n = len(g), len(g[0])
        dis = [[float('inf')] * n for _ in range(m)]
        h = [(0, 0, 0)]
        dis[0][0] = 0
        heapq.heapify(h)
        while h:
            times, x, y = heapq.heappop(h)
            if x == m - 1 and y == n - 1: return times
            if times > dis[x][y]: continue
            for dx, dy in d:
                nx, ny = dx + x, dy + y
                if 0 <= nx < m and 0 <= ny < n:
                    nxt_time = max(g[nx][ny], times) + (x + y) % 2 + 1
                    if nxt_time < dis[nx][ny]:
                        dis[nx][ny] = nxt_time
                        heapq.heappush(h, (nxt_time, nx, ny))

# @lc code=end

